src/System.Threading.Timer/tests/TimerFiringTests.cs

```diff
@@ -0,0 +1,226 @@
+        {
+            Assert.Same(s, state);
+        }), state, TimeSpan.FromMilliseconds(100), TimeSpan.FromMilliseconds(Timeout.Infinite) /* not relevant */)) { }
+    }
+
+    [Fact]
+    public void Timer_Fires_AndPassesNullStateThroughCallback()
+    {
+        using (var t = new Timer(new TimerCallback((object s) =>
+        {
+            Assert.Null(s);
+        }), null, TimeSpan.FromMilliseconds(100), TimeSpan.FromMilliseconds(Timeout.Infinite) /* not relevant */)) { }
+    }
+
+    [Fact]
+    public void Timer_Fires_After_DueTime_AndOn_Period()
+    {
+        int count = 0;
+        AutoResetEvent are = new AutoResetEvent(false);
+
+        using (var t = new Timer(new TimerCallback((object s) =>
+        {
+            count++;
+            if (count >= 2)
+            {
+                are.Set();
+            }
+        }), null, TimeSpan.FromMilliseconds(250), TimeSpan.FromMilliseconds(50)))
+        {
+            Assert.True(are.WaitOne(TimeSpan.FromMilliseconds(2000)));
+        }
+    }
+
+    [Fact]
+    public void Timer_FiresOnlyOnce_OnDueTime_With_InfinitePeriod()
+    {
+        int count = 0;
+        AutoResetEvent are = new AutoResetEvent(false);
+
+        using (var t = new Timer(new TimerCallback((object s) =>
+        {
+            count++;
+            if (count >= 2)
+            {
+                are.Set();
+            }
+        }), null, TimeSpan.FromMilliseconds(100), TimeSpan.FromMilliseconds(Timeout.Infinite) /* not relevant */))
+        {
+            Assert.False(are.WaitOne(TimeSpan.FromSeconds(1)));
+        }
+    }
+
+    [Fact]
+    public void MultipleTimers_ShouldFire_InCorrectOrder()
+    {
+        Object t1State = new Object();
+        Object t2State = new Object();
+        Timer t1 = null;
+        Timer t2 = null;
+        int count = 0;
+        AutoResetEvent are = new AutoResetEvent(false);
+        TimerCallback tc = new TimerCallback((object o) =>
+        {
+            if (count == 0)
+            {
+                Assert.Same(t2State, o);
+                count++;
+            }
+            else
+            {
+                Assert.Same(t1State, o);
+                are.Set();
+            }
+        });
+
+        using (t1 = new Timer(tc, t1State, TimeSpan.FromSeconds(1), TimeSpan.FromMilliseconds(-1)))
+        using (t2 = new Timer(tc, t2State, TimeSpan.FromMilliseconds(0), TimeSpan.FromMilliseconds(-1)))
+        {
+            // Wait for both events to fire
+            Assert.True(are.WaitOne(TimeSpan.FromMilliseconds(1500 /*1.5 seconds*/)));
+        }
+    }
+
+    [Fact]
+    public void Timer_CanDisposeSelfInCallback()
+    {
+        // There's nothing to validate besides that we don't throw an exception, so rely on xunit 
+        // to catch any exception that would be thrown and signal a test failure
+        Timer t = null;
+        TimerCallback tc = new TimerCallback((object o) => t.Dispose());
+        t = new Timer(tc, null, 100, -1);
```

Comment:
You still need to wait for the callback to be invoked.  The tests could have ended by the time the callback eventually runs, and the callback is happening on a thread pool thread xunit has no knowledge of.  Plus, it's quite likely that the Timer will get GC'd and it'll never fire.